src/crypto/crypto.h

```diff
@@ -125,6 +125,26 @@ namespace crypto {
   };
   using x25519_secret_key = epee::mlocked<tools::scrubbed<x25519_secret_key_>>;
 
+  union alignas(size_t) generic_public_key
+  {
+    ed25519_public_key ed25519;
+    public_key         monero;
+    unsigned char      data[sizeof(ed25519_public_key)];
+    static constexpr generic_public_key null() { return {0}; }
+    operator bool() const { return memcmp(data, null().data, sizeof(data)); }
+  };
```

Comment:
> because the codegen for variant sucks.

Can you elaborate?  AFAIK most of the codegen issues with variant are around visiting, which you wouldn't need to use here (just `get_if`s for the two types).

> With the variant method you also decouple the type from the data for de/serialisation?

Sure, but you *always* do that with serialization, deserialization should pick it up again (which means the serialization needs to carry the type as well).

Reply:
> Can you elaborate? AFAIK most of the codegen issues with variant are around visiting, which you wouldn't need to use here (just get_ifs for the two types).

Yes. You're somewhat right its good at -O2 for the std implementation but boost is just crazy, https://godbolt.org/z/awP_wd

> Sure, but you always do that with serialization, deserialization should pick it up again (which means the serialization needs to carry the type as well).
Yeah actually not a big deal after some thought as it can easily be achieved 

```cpp
struct generic_key 
{
    variant<a, b> key;
    bool is_a;
};

Reply:
Loki-mq is using mapbox::variant because most things in boost are too heavy.